File: apps/web/src/api/planWorkflow.ts
import type {
  PlanSummaryDTO,
  DailyPlanTaskDTO,
  CreateDailyPlanTaskDTO,
  UpdateDailyPlanTaskDTO,
} from "../types/planWorkflow.dto"

export class ApiError extends Error {
  status: number

  constructor(status: number, message: string) {
    super(message)
    this.name = "ApiError"
    this.status = status
  }
}

async function handleResponse<T>(res: Response): Promise<T> {
  if (!res.ok) {
    const text = await res.text()
    throw new ApiError(res.status, text || `Request failed with status ${res.status}`)
  }
  return res.json()
}

// Helper function to extract auth token
function getAccessToken(): string | null {
  try {
    const raw = localStorage.getItem('learnsmart-tokens')
    if (!raw) return null
    return JSON.parse(raw).access_token
  } catch (err) {
    return null
  }
}

export async function getPlanSummary(): Promise<PlanSummaryDTO> {
  const token = getAccessToken()
  const res = await fetch(`/api/plan-workflow/summary`, {
    headers: token ? { "Authorization": `Bearer ${token}` } : {},
    credentials: "include",
  })
  return handleResponse<PlanSummaryDTO>(res)
}

export async function getDailyPlanTasks(date?: string): Promise<DailyPlanTaskDTO[]> {
  const token = getAccessToken()
  const query = date ? `?date=${encodeURIComponent(date)}` : ""
  const res = await fetch(`/api/plan-workflow/daily-tasks${query}`, {
    headers: token ? { "Authorization": `Bearer ${token}` } : {},
    credentials: "include",
  })
  return handleResponse<DailyPlanTaskDTO[]>(res)
}

export async function createDailyTask(payload: CreateDailyPlanTaskDTO, date?: string): Promise<DailyPlanTaskDTO> {
  const token = getAccessToken()
  const headers: Record<string, string> = { "Content-Type": "application/json" }
  if (token) headers["Authorization"] = `Bearer ${token}`
  const query = date ? `?date=${encodeURIComponent(date)}` : ""
  const res = await fetch(`/api/plan-workflow/daily-tasks${query}`, {
    method: "POST",
    headers,
    credentials: "include",
    body: JSON.stringify(payload),
  })
  return handleResponse<DailyPlanTaskDTO>(res)
}

export async function updateDailyTask(
  taskId: string,
  payload: UpdateDailyPlanTaskDTO
): Promise<DailyPlanTaskDTO> {
  const token = getAccessToken()
  const headers: Record<string, string> = { "Content-Type": "application/json" }
  if (token) headers["Authorization"] = `Bearer ${token}`

  const res = await fetch(`/api/plan-workflow/daily-tasks/${taskId}`, {
    method: "PATCH",
    headers,
    credentials: "include",
    body: JSON.stringify(payload),
  })
  return handleResponse<DailyPlanTaskDTO>(res)
}

export async function deleteDailyTask(taskId: string): Promise<void> {
  const token = getAccessToken()
  const res = await fetch(`/api/plan-workflow/daily-tasks/${taskId}`, {
    method: "DELETE",
    headers: token ? { "Authorization": `Bearer ${token}` } : {},
    credentials: "include",
  })
  await handleResponse<void>(res)
}

export async function generateWeeklyPlan(startDate?: string): Promise<{ weekTemplate: Array<Record<string, any>> }> {
  const token = getAccessToken()
  const query = startDate ? `?start_date=${encodeURIComponent(startDate)}` : ""
  const res = await fetch(`/api/plan-workflow/weekly-plan/generate${query}`, {
    method: "POST",
    headers: token ? { "Authorization": `Bearer ${token}` } : {},
    credentials: "include",
  })
  return handleResponse<{ weekTemplate: Array<Record<string, any>> }>(res)
}

File: backend/app/api/planWorkflow.py
from fastapi import APIRouter, Depends, HTTPException, Query, Path
from typing import Optional

from app.core.dependencies import get_current_user, get_postgres
from app.services.planWorkflow_service import PlanWorkflowService
from app.models.planWorkflow import CreateDailyPlanTaskDTO, UpdateDailyPlanTaskDTO

router = APIRouter(prefix="/plan-workflow", tags=["Plan Workflow"])

@router.get("/summary")
async def get_plan_summary(
    current_user=Depends(get_current_user),
    db=Depends(get_postgres)
):
    user_id = str(current_user["id"])
    service = PlanWorkflowService(db)
    try:
        return await service.get_plan_summary(user_id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/daily-tasks")
async def get_daily_tasks(
    date: Optional[str] = Query(None),
    current_user=Depends(get_current_user),
    db=Depends(get_postgres)
):
    user_id = str(current_user["id"])
    service = PlanWorkflowService(db)
    try:
        return await service.get_daily_tasks(user_id, date)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/daily-tasks")
async def create_daily_task(
    payload: CreateDailyPlanTaskDTO,
    date: Optional[str] = Query(None),
    current_user=Depends(get_current_user),
    db=Depends(get_postgres)
):
    user_id = str(current_user["id"])
    service = PlanWorkflowService(db)
    try:
        return await service.create_daily_task(user_id, payload, date)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.patch("/daily-tasks/{task_id}")
async def update_daily_task(
    payload: UpdateDailyPlanTaskDTO,
    task_id: str = Path(...),
    current_user=Depends(get_current_user),
    db=Depends(get_postgres)
):
    user_id = str(current_user["id"])
    service = PlanWorkflowService(db)
    try:
        return await service.update_daily_task(user_id, task_id, payload)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.delete("/daily-tasks/{task_id}")
async def delete_daily_task(
    task_id: str = Path(...),
    current_user=Depends(get_current_user),
    db=Depends(get_postgres)
):
    user_id = str(current_user["id"])
    service = PlanWorkflowService(db)
    try:
        success = await service.delete_daily_task(user_id, task_id)
        if success:
            return {"status": "ok", "message": "Task deleted successfully"}
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/daily-tasks/{task_id}/complete")
async def complete_daily_task(
    task_id: str = Path(...),
    current_user=Depends(get_current_user),
    db=Depends(get_postgres)
):
    user_id = str(current_user["id"])
    service = PlanWorkflowService(db)
    try:
        return await service.mark_task_completed(user_id, task_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/weekly-plan/refresh")
async def refresh_weekly_plan(
    start_date: Optional[str] = Query(None),
    current_user=Depends(get_current_user),
    db=Depends(get_postgres)
):
    user_id = str(current_user["id"])
    service = PlanWorkflowService(db)
    try:
        return {"weekTemplate": await service.refresh_weekly_plan(user_id, start_date)}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/weekly-plan/generate")
async def generate_weekly_plan(
    start_date: Optional[str] = Query(None),
    current_user=Depends(get_current_user),
    db=Depends(get_postgres)
):
    user_id = str(current_user["id"])
    service = PlanWorkflowService(db)
    try:
        return {"weekTemplate": await service.generate_weekly_plan(user_id, start_date)}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
